File: CameraCalibration/cameraCalibrationWithUndistortion.py
#!/usr/bin/env python3
"""Calibrate a camera and compare two OpenCV undistortion workflows."""

from __future__ import annotations

# argparse documents the inputs, outputs, headless mode, and validation mode.
import argparse
# sys supplies standard error and the final process status.
import sys
# Path makes input and output locations explicit and platform independent.
from pathlib import Path
# Sequence permits direct testing without modifying process-global arguments.
from typing import Sequence

# Shared calibration logic prevents the two lessons from drifting apart.
from calibration_utils import (
    calibrate_images,
    print_calibration,
    undistort_sample,
    validate_calibration,
)


# Locate bundled data relative to the source file, not the current directory.
PROJECT_DIR = Path(__file__).resolve().parent
# The wildcard selects all 41 tracked checkerboard images.
DEFAULT_IMAGE_PATTERN = str(PROJECT_DIR / "images" / "*.jpg")
# The two documented methods should remain below one mean intensity level; the
# tighter 0.1 threshold passed on both exact reference versions.
MAXIMUM_METHOD_DIFFERENCE = 0.1


def parse_args(arguments: Sequence[str] | None = None) -> argparse.Namespace:
    """Parse command-line options, optionally from a test-provided sequence."""

    # Use the module documentation as the command's concise help text.
    parser = argparse.ArgumentParser(description=__doc__)
    # A quoted glob can select an alternate same-sized calibration dataset.
    parser.add_argument(
        "--images",
        default=DEFAULT_IMAGE_PATTERN,
        help="Calibration image glob (default: bundled images)",
    )
    # By default the first sorted calibration image demonstrates undistortion.
    parser.add_argument(
        "--sample",
        type=Path,
        help="Image to undistort (default: first calibration image)",
    )
    # Keep generated files out of the source tree during normal validation.
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=Path.cwd(),
        help="Output directory (default: current directory)",
    )
    # Headless mode avoids imshow and waitKey on servers and in CI.
    parser.add_argument(
        "--no-display",
        action="store_true",
        help="Skip interactive checkerboard and undistortion windows",
    )
    # Validation checks both calibration quality and method agreement.
    parser.add_argument(
        "--validate",
        action="store_true",
        help="Check calibration and undistortion invariants",
    )
    # argparse reads sys.argv when arguments is None.
    return parser.parse_args(arguments)


def run(options: argparse.Namespace) -> None:
    """Execute calibration and both undistortion methods."""

    # Calibrate from the complete selected checkerboard set.
    calibration = calibrate_images(
        options.images,
        display=not options.no_display,
    )
    # Report the same reusable camera parameters as the calibration-only lesson.
    print_calibration(calibration)
    # A user-supplied sample wins; otherwise sorting makes the first match stable.
    sample_path = options.sample or calibration.image_paths[0]
    # Produce one image with direct undistortion and one with precomputed maps.
    _, _, method_difference = undistort_sample(
        calibration,
        sample_path,
        options.output_dir,
        display=not options.no_display,
    )
    # Print enough precision to expose the version-specific interpolation change.
    print(
        "Undistortion method mean absolute difference: "
        f"{method_difference:.9f}"
    )

    # Custom exploratory runs need not meet the bundled fixture's validation.
    if options.validate:
        # Check the calibration result before trusting it for rectification.
        validate_calibration(calibration)
        # The two documented workflows should remain visually equivalent.
        if method_difference >= MAXIMUM_METHOD_DIFFERENCE:
            raise RuntimeError(
                "Direct undistortion and remapping differ by more than expected"
            )
        # Tests require an unambiguous marker after every check succeeds.
        print("Validation passed")


def main(arguments: Sequence[str] | None = None) -> int:
    """Run the CLI and convert operational failures into a clean exit status."""

    try:
        # Let argparse keep its standard help and malformed-option behavior.
        options = parse_args(arguments)
        # Exercise the same run function used by direct imports and tests.
        run(options)
    except Exception as error:
        # imread, imwrite, OpenCV, and validation failures share one clear format.
        print(f"Error: {error}", file=sys.stderr)
        return 1
    # Zero communicates successful image generation and optional validation.
    return 0


if __name__ == "__main__":
    # SystemExit exposes main's status to shells, CTest, and subprocess tests.
    raise SystemExit(main())

File: CameraCalibration/cameraCalibrationWithUndistortion.cpp
// Shared helpers keep both tutorial executables numerically identical.
#include "calibration_utils.hpp"

// Filesystem resolves the bundled wildcard and explicit output directory.
#include <filesystem>
// The CLI prints help, results, validation markers, and errors.
#include <iostream>
// Invalid options and failed invariants use standard exceptions.
#include <stdexcept>
// Strings store option tokens and the calibration image pattern.
#include <string>

// Keep path-heavy option handling concise.
namespace fs = std::filesystem;

// Both exact reference versions remain below this mean intensity difference.
constexpr double kMaximumMethodDifference = 0.1;

// Hold command-line choices separately from calibration and image results.
struct Options {
    // CMake supplies an absolute bundled-asset directory at compile time.
    std::string images =
        (fs::path(TUTORIAL_IMAGE_DIR) / "*.jpg").string();
    // An empty sample means "use the first sorted calibration image."
    fs::path sample;
    // Generated images go to the caller's current directory by default.
    fs::path output_dir = fs::current_path();
    // Preserve the original interactive lesson unless headless mode is selected.
    bool display = true;
    // Validation remains optional for custom datasets.
    bool validate = false;
    // Help exits without reading images or creating output.
    bool show_help = false;
};

// Print the exact interface accepted by this executable.
void print_usage(const char* executable) {
    std::cout
        << "Usage: " << executable
        << " [--images GLOB] [--sample IMAGE] [--output-dir PATH]\n"
        << "       [--no-display] [--validate]\n"
        << "\n"
        << "Options:\n"
        << "  --images GLOB     Calibration image pattern (default: bundled images)\n"
        << "  --sample IMAGE    Image to undistort (default: first calibration image)\n"
        << "  --output-dir PATH Output directory (default: current directory)\n"
        << "  --no-display      Skip all interactive windows\n"
        << "  --validate        Check calibration and undistortion invariants\n"
        << "  -h, --help        Show this help text\n";
}

// Convert process arguments into a validated Options value.
Options parse_options(const int argc, char** argv) {
    // Begin with safe, documented defaults.
    Options options;
    // argv[0] names the executable, so start parsing at index one.
    for (int index = 1; index < argc; ++index) {
        // Copy the token before comparing it with supported option names.
        const std::string argument = argv[index];
        // All three path options require exactly one following value.
        if ((argument == "--images" ||
             argument == "--sample" ||
             argument == "--output-dir") &&
            index + 1 >= argc) {
            throw std::invalid_argument(argument + " requires a value");
        }

        if (argument == "--images") {
            // Increment first so the wildcard is not parsed again as an option.
            options.images = argv[++index];
        } else if (argument == "--sample") {
            // The sample must have the calibration image resolution.
            options.sample = argv[++index];
        } else if (argument == "--output-dir") {
            // Nested output directories are created by the shared helper.
            options.output_dir = argv[++index];
        } else if (argument == "--no-display") {
            // Headless mode prevents every HighGUI call.
            options.display = false;
        } else if (argument == "--validate") {
            // Validate the camera model and agreement between both methods.
            options.validate = true;
        } else if (argument == "-h" || argument == "--help") {
            // Defer the successful exit until main can print usage.
            options.show_help = true;
        } else {
            // Reject option typos instead of silently running with defaults.
            throw std::invalid_argument("Unknown argument: " + argument);
        }
    }
    // Return one complete configuration to main.
    return options;
}

// Run the calibration-plus-undistortion lesson.
int main(const int argc, char** argv) {
    try {
        // Resolve and validate all requested behavior before touching files.
        Options options = parse_options(argc, argv);
        // Help is successful and deliberately side-effect free.
        if (options.show_help) {
            print_usage(argv[0]);
            return 0;
        }

        // Estimate intrinsics, distortion, and checkerboard poses.
        const tutorial::CalibrationResult calibration =
            tutorial::calibrate_images(options.images, options.display);
        // Report the same camera values as the calibration-only lesson.
        tutorial::print_calibration(calibration);

        // Sorting makes the first calibration image a deterministic default.
        if (options.sample.empty()) {
            options.sample = calibration.image_paths.front();
        }
        // Generate direct and precomputed-map undistortion results.
        const tutorial::UndistortionResult undistortion =
            tutorial::undistort_sample(
                calibration,
                options.sample,
                options.output_dir,
                options.display
            );
        // Print enough precision to expose the OpenCV 5 interpolation difference.
        std::cout
            << "Undistortion method mean absolute difference: "
            << undistortion.mean_absolute_difference << '\n';

        // Custom exploratory runs need not meet the bundled validation contract.
        if (options.validate) {
            // Reject malformed or poor-quality camera parameters first.
            tutorial::validate_calibration(calibration);
            // Check matrices, method agreement, and both decoded output files.
            tutorial::validate_undistortion(
                undistortion,
                kMaximumMethodDifference
            );
            // CTest requires this marker only after every check succeeds.
            std::cout << "Validation passed\n";
        }
        // Zero signals that calibration, output, and requested checks succeeded.
        return 0;
    } catch (const std::exception& error) {
        // Convert option, image, OpenCV, filesystem, and validation failures.
        std::cerr << "Error: " << error.what() << '\n';
        return 1;
    }
}
